### ISSUE TITLE
problem with :revision keyword

### ISSUE THREAD

**Issue by lemzwerg** [Homebrew/homebrew#44981]
Put [this formula] (https://raw.githubusercontent.com/lemzwerg/homebrew/29196533fa50af4a90e875c8ca554e1638127639/Library/Formula/ttfautohint-1.3.56.rb) into `Library/Formula`, which is a slight variation of the already present `ttfautohint.rb` file.  Then say

    brew install ttfautohint-1.3.56

`ttfautohint --version` should contain 1.3.56.  However, instead of picking up the selected git revision as specified with the `url` keyword, homebrew uses git HEAD instead.  If I follow the above command with

      brew reinstall ttfautohint-1.3.56

I finally get the right version.

In other words, it seems that Homebrew doesn't handle the `:revision` keyword correctly if the git repository to be accessed is not yet cloned.

**Comment by mikemcquaid** [comment]
Can you provide a Gist of all the commands you've run, their output and explain the expected and actual behaviour when they differ? Thanks!

**Comment by lemzwerg** [comment]
With 'Gist' you mean the output as produced with homebrew's `--verbose` flag?  Will do as soon as I'm back to my Mac.  However, the commands I've used are exactly as above, and nothing else (after updating homebrew, of course).  The first command misses a proper call to `git checkout`, while the second one does the right thing.

**Comment by mikemcquaid** [comment]
I mean paste it to https://gist.github.com

**Comment by apjanke** [comment]
I can reproduce. https://gist.github.com/apjanke/671b394a34c5be4fce2c

Short story:

```ruby
$ rm -rf /Library/Caches/Homebrew/ttfautohint-1.3.56--git
$ brew install ttfautohint-1.3.56
$ ttfautohint-1.3.56 --version | head -1
ttfautohint 1.4.2-a5e4
$ brew reinstall ttfautohint-1.3.56
==> Reinstalling ttfautohint-1.3.56
$ ttfautohint-1.3.56 --version | head -1
ttfautohint 1.3.56-d99a
```

I think I see the problem. In [download_strategy.rb](https://github.com/Homebrew/homebrew/blob/e646a9ca2dd875d35596ae3383e7288e20ec8280/Library/Homebrew/download_strategy.rb),  `GitDownloadStrategy` only calls `checkout` inside `update`, not in `clone_repo`. So the revision only gets checked out if you already have the repo in your cache and just update it. The initial download and build attempt of a Git repo using just `:revision` will get the wrong `:revision`.

The `stage()` method in `VCSDownloadStrategy` is what's producing that "Checking out revision..." output, but the stage methods don't actually check anything out; they just copy the checked-out repo contents to the build directory. It's lying about what it's doing. And the check for the correct checked-out revision is skipped, because it's only done if :tag` is provided.

```
class VCSDownloadStrategy < AbstractDownloadStrategy
  def fetch
    ohai "Cloning #{@url}"

    if cached_location.exist? && repo_valid?
      puts "Updating #{cached_location}"
      update
    elsif cached_location.exist?
      puts "Removing invalid repository from cache"
      clear_cache
      clone_repo
    else
      clone_repo
    end

    if @ref_type == :tag && @revision && current_revision
      unless current_revision == @revision
        raise <<-EOS.undent
          #{@ref} tag should be #{@revision}
          but is actually #{current_revision}
        EOS
      end
    end
  end

  def stage
    ohai "Checking out #{@ref_type} #{@ref}" if @ref_type && @ref
  end

end

class GitDownloadStrategy < VCSDownloadStrategy

  def clone_repo
    safe_system "git", *clone_args
    cached_location.cd do
      safe_system "git", "config", "homebrew.cacheversion", cache_version
      update_submodules if submodules?
    end
  end

  def update
    cached_location.cd do
      config_repo
      update_repo
      checkout
      reset
      update_submodules if submodules?
    end
  end

  def stage
    super
    cp_r File.join(cached_location, "."), Dir.pwd
  end

end
```

In the `MercurialDownloadStrategy`, the revision checkout is done in both the initial `clone_repo` and subsequent updates, as `update` is just aliased to `clone_repo`.

Sounds like `GitDownloadStrategy` should do a `checkout` in its `clone_repo`, too.

You can see the issue cropping up with `colortail`, too. The first time I run `brew install --build-from-source colortail`, the build will fail at the patching stage. But if I immediately run `brew install --build-from-source colortail` again, the patch will succeed and the build works.

**Comment by mikemcquaid** [comment]
Nice debugging @apjanke :clap:. Could you try and open a pull request? This should help and I'm happy to walk you through anything else: https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/How-To-Open-a-Homebrew-Pull-Request-(and-get-it-merged).md#how-to-open-a-homebrew-pull-request-and-get-it-merged

Thanks!

**Comment by apjanke** [comment]
Here's a PR: #45019

**Comment by apjanke** [comment]
@lemzwerg: The PR to fix this has been mreged in. Could you `brew update` and check to see if it resolved your issue?

**Comment by lemzwerg** [comment]
It works fine now.  Thanks for the quick fix!
